src/vm/eventpipebuffermanager.cpp

```diff
@@ -353,14 +353,12 @@ bool EventPipeBufferManager::WriteEvent(Thread *pThread, EventPipeSession &sessi
         }
         else
         {
-            // The event is still enabled.  Mark that the thread is now writing an event.
-            pThreadBufferList->SetThreadEventWriteInProgress(true);
-
-            // Attempt to write the event to the buffer.  If this fails, we should allocate a new buffer.
-            allocNewBuffer = !pBuffer->WriteEvent(pEventThread, session, event, payload, pActivityId, pRelatedActivityId, pStack);
+            {
+                SpinLockHolder _slh(pThreadBufferList->GetLock());
 
-            // Mark that the thread is no longer writing an event.
-            pThreadBufferList->SetThreadEventWriteInProgress(false);
+                // Attempt to write the event to the buffer.  If this fails, we should allocate a new buffer.
+                allocNewBuffer = !pBuffer->WriteEvent(pEventThread, session, event, payload, pActivityId, pRelatedActivityId, pStack);
```

Comment:
I just realized that this call could lead to doing a full stackwalk - we need to figure out a way to make the spin lock region much smaller than that.

Reply:
This is still work in progress, with #23185, we were able to move stack walk out of the critical section. 

We still have a couple of issues remaining:

- [ ] SpinLock nesting - right now I just commented out the assert to allow SpinLock nest, which is not good.
- [ ] SpinLock being disposed of while we are still holding the lock in the DeAllocate() code path

The plan to solve the SpinLock nesting problem is to turn EventPipeBufferManager::m_lock into a critical section, and the plan to solve the SpinLock being disposed problem is to store the SpinLock itself somewhere else and never dispose it.